src/main/java/org/truffleruby/core/inlined/InlinedIsNilNode.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2017 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 1.0
+ * GNU General Public License version 2
+ * GNU Lesser General Public License version 2.1
+ */
+package org.truffleruby.core.inlined;
+
+import org.truffleruby.RubyContext;
+import org.truffleruby.language.dispatch.RubyCallNodeParameters;
+import org.truffleruby.language.methods.LookupMethodNode;
+
+import com.oracle.truffle.api.dsl.Cached;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+
+public abstract class InlinedIsNilNode extends UnaryInlinedOperationNode {
+
+    protected static final String METHOD = "nil?";
+
+    public InlinedIsNilNode(RubyContext context, RubyCallNodeParameters callNodeParameters) {
+        super(callNodeParameters,
+                context.getCoreMethods().nilClassIsNilAssumption);
+    }
+
+    @Specialization(guards = "isNil(self)", assumptions = "assumptions")
+    boolean nil(Object self) {
+        return true;
+    }
+
+    @Specialization(guards = {
+            "!isForeignObject(self)",
+            "lookupNode.lookup(frame, self, METHOD) == coreMethods().KERNEL_IS_NIL",
```

Comment:
I take it this lookup is done to ensure the `nil?` method hasn't been redefined? If so, a short comment might be helpful since this deviates a bit from the typical assumption check we do for inlined nodes.

Reply:
We need to make sure `nil?` on that object resolves to the original `Kernel#nil?`.
This is the same in other inlined nodes like `block_given?`, I think it's fairly clear by the code itself.
The only way I think of  to avoid that would be to have a global assumption per method name redefined (so any definition of a `nil?` method would invalidate it), but it seems very fragile and a lookup is just a Shape check.